smac/tae/base.py

```diff
@@ -0,0 +1,316 @@
+    3- This interface also offers the method wait() as a mechanism to make
+       sure we have enough data in the next iteration to make a decision. For
+       example, the intensifier might not be able to select the next challenger
+       until more results are available.
+
+
+    Attributes
+    ----------
+
+    results
+    ta
+    stats
+    run_obj
+    par_factor
+    cost_for_crash
+    abort_first_run_crash
+
+    Parameters
+    ---------
+    ta : list
+        target algorithm command line as list of arguments
+    stats: Stats()
+         stats object to collect statistics about runtime and so on
+    run_obj: str
+        run objective of SMAC
+    par_factor: int
+        penalization factor
+    cost_for_crash : float
+        cost that is used in case of crashed runs (including runs
+        that returned NaN or inf)
+    abort_on_first_run_crash: bool
+        if true and first run crashes, raise FirstRunCrashedException
+    """
+    def __init__(
+        self,
+        ta: typing.Union[typing.List[str], typing.Callable],
+        stats: Stats,
+        run_obj: str = "runtime",
+        par_factor: int = 1,
+        cost_for_crash: float = float(MAXINT),
+        abort_on_first_run_crash: bool = True,
+    ):
+
+        # The results is a FIFO structure, implemented via a list
+        # (because the Queue lock is not pickable). Finished runs are
+        # put in this list and collected via process_finished_runs
+        self.results = []  # type: typing.List[typing.Tuple[RunInfo, RunValue]]
+
+        # Below state the support for a Runner algorithm that
+        # implements a ta
+        self.ta = ta
+        self.stats = stats
+        self.run_obj = run_obj
+        self.par_factor = par_factor
+        self.cost_for_crash = cost_for_crash
+        self.abort_on_first_run_crash = abort_on_first_run_crash
+        self.logger = PickableLoggerAdapter(
+            self.__module__ + '.' + self.__class__.__name__)
+        self._supports_memory_limit = False
+
+        super().__init__()
+
+    @abstractmethod
+    def submit_run(self, run_info: RunInfo) -> None:
+        """This function submits a configuration
+        embedded in a RunInfo object, and uses one of the workers
+        to produce a result (such result will eventually be available
+        on the self.results FIFO).
+
+        This interface method will be called by SMBO, with the expectation
+        that a function will be executed by a worker.
+
+        What will be executed is dictated by run_info, and "how" will it be
+        executed is decided via the child class that implements a run() method.
+
+        Because config submission can be a serial/parallel endeavor,
+        it is expected to be implemented by a child class.
+
+        Parameters
+        ----------
+        run_info: RunInfo
+            An object containing the configuration and the necessary data to run it
+
+        """
+        pass
+
+    def run(
```

Comment:
yeah (I didn;t make it originally an abstract class so that I don't copy the same code to serial/parallel, but I just changed it because seems confusing)!